utils/s3/client.cc

```diff
@@ -1092,6 +1093,117 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
     return data_sink(std::make_unique<upload_jumbo_sink>(shared_from_this(), std::move(object_name), max_parts_per_piece, as));
 }
 
+class client::chunked_download_source final : public seastar::data_source_impl {
+    shared_ptr<client> _client;
+    sstring _object_name;
+    seastar::abort_source* _as;
+    range _range;
+    static constexpr size_t _max_buffers_size = 5_MiB;
+    std::deque<temporary_buffer<char>> _buffers;
+    size_t _buffers_size = 0;
+    semaphore _buffers_sem;
+    bool _is_closing = false;
+    bool _is_done = false;
+    seastar::condition_variable _buffer_cv;
+    future<> _filling_fiber = make_ready_future<>();
+
+    future<> make_filling_fiber() {
+        while (true) {
+            try {
+                s3l.trace("Fiber starts cycle for object '{}'", _object_name);
+                if ((_as && _as->abort_requested()) || _is_closing) {
```

Comment:
nit: Checking _as is excessive. The download_source would abort is .get() with aborted exception if it happens